lib/tapioca/runtime/reflection.rb

```diff
@@ -174,14 +175,24 @@ def resolve_loc(locations)
         resolved_loc.absolute_path || ""
       end
 
-      sig { params(singleton_class: Module).returns(T.nilable(Module)) }
-      def attached_class_of(singleton_class)
-        # https://stackoverflow.com/a/36622320/98634
-        result = ObjectSpace.each_object(singleton_class).find do |klass|
-          singleton_class_of(T.cast(klass, Module)) == singleton_class
+      # On Ruby 3.2 or newer, Class defines an attached_object method that returns the
+      # attached class of a singleton class without iterating ObjectSpace. On older
+      # versions of Ruby, we fall back to iterating ObjectSpace.
+      if CLASS_ATTACHED_OBJECT
+        sig { params(singleton_class: Class).returns(T.nilable(Module)) }
+        def attached_class_of(singleton_class)
+          T.cast(singleton_class.attached_object, Module)
```

Comment:
So this code is not always going to work. The method name was chosen as `attached_object` for the reason that while a singleton class will always be a class, the object that has that singleton class does not need to be a module or a class. It can be a simple object. For example:
```ruby
foo = "foo"
sc = foo.singleton_class
Reflection.attached_class_of(sc)
```
should return `nil` since `sc` is not attached to a class, it is attached to a string object.

The fix is:
```suggestion
          result = singleton_class.attached_object
          if Module === result
            result
          else
            nil
          end
```